hadoop-hdds/docs/content/design/container-reconciliation.md

```diff
@@ -0,0 +1,246 @@
+
+The only extra information we will need to store is the container merkle tree on each datanode container replica. The current proposal is store this separately as a proto file on disk so that it can be copied over the network exactly as stored. The structure would look something like this (not finalized, for illustrative purposes only):
+
+```
+ContainerChecksum: {
+  algorithm: CRC32
+  checksum: 12345
+  repeated BlockChecksum
+}
+BlockChecksum: {
+  algorithm: CRC32
+  checksum: 12345
+  length: 5
+  deleted: false
+  healthy: true
+  repeated ChunkChecksum
+}
+ChunkChecksum: {
+  algorithm: CRC32
+  checksum: 12345
+  length: 5
+  offset: 5
+  healthy: true
+}
+```
+
+- The `deleted` flag in the block proto allows it to serve as a tombstone entry so that block deletion does not affect the container level hash.
+
+- The `algorithm` field is the algorithm used to aggregate hashes from the layer below which results in that object's current hash.
+    - We can define an order of chunks and blocks based on their IDs so that the hash does not need to be commutative (order independent).
+
+- The `healthy` field in the chunk checksum being true indicates that the computed chunk checksum used to fill in the merkle tree matches the chunk checksum stored in RocksDB.
+
+- The `healthy` field in the block checksum being true indicates that all chunks within the block are marked as `healthy = true`.
+
+- We may be able to re-use the `ChunkInfo` and `ChecksumData` messages in `DatanodeClientProtocol.proto` for chunk information and checksums. Other structures in the merkle tree storage will require new objects to be defined.
+
+## APIs
+
+The following APIs would be added to datanodes to support container reconciliation. The actions performed when calling them is defined in [Events](#events).
+
+- `reconcileContainer(containerID, List<Replica>)`
+    - Instructs a datanode to reconcile its copy of the specified container with the provided list of replica's containers.
+    - Datanodes will call `getContainerHashes` for each replica to identify repairs needed, and use existing chunk read/write APIs to do repairs necessary.
+    - This would likely be a new command as part of the SCM heartbeat protocol, not actually a new API.
+- `getContainerHashes(containerID)`
+    - A datanode API that returns the merkle tree for a given container. The proto structure would be similar to that outlined in [Storage](#storage)
+
+## Events
+
+This section defines how container reconciliation will function as events occur in the system.
+
+### Reconiliation Events
+
+ These events occur as part of container reconciliation in the happy path case.
+
+#### On Data Write
+
+- No change to existing datanode operations. Existing chunk checksums passed and validated from the client will be persisted to RocksDB.
+
+#### On Container Close
+
+- Container checksum is calculated synchonously from existing checksums. This calculation must finish before the close is acked back to SCM.
```

Comment:
We can, it would be best to do this once we can measure the performance hit of doing it as part of write or for an open container.

Reply:
This will be tricky to implement, because multiple blocks can be written to a container in parallel and the checksum will require the same order across all nodes. Block IDs are handed out to clients in an increasing order, but there is no guarantee that the client(s) write them to the container in that order.